### PR TITLE
Serve EmberCLI tests as mountable engine

### LINKED ISSUE #18: Make original ember tests available through rails routes
Make http://localhost:4200/tests/* available through rails somehow.

//cc @mwoods79

### PR DESCRIPTION
**Pull request by seanpdoyle**
Closes #18

Expose EmberCLI test suite through Rails

### CODE REVIEW

**Comment by seanpdoyle** [comment]
Unfortunately, this doesn't yet work.

The `app/frontend/tests/index.html` is being served properly from the `EmberCLI` dist_path, but the script and link tags are pointing to assets like `assets/vendor.css`, instead of `assets/frontend/vendor-FINGERPRINT.css`.

Any ideas? cc: @rwjblue @rwz @rondale-sc

**Comment by rwz** [comment]
This is amazing. Thanks for this great work!

We could maybe perform a naive gsub before serving the html to replace `assets/` with `assets/:app_name/`. That's easy and solves the problem.

**Comment by rwz** [comment]
Also, why remove `railties` from dependiencies? `Rails::Engine` is a part of `railties` as well AFAIK.

File: lib/ember-cli-rails.rb [hunk 1]
```diff
@@ -1,4 +1,4 @@
-require "ember-cli/railtie" if defined?(Rails)
+require "ember-cli/engine" if defined?(Rails)
```
Comment (seanpdoyle) [inline lib/ember-cli-rails.rb]: @rwz if the railties / engine is only loaded when `Rails` constant exists, we're really dependent on Rails, not railties.

If Rails is included in the dependencies (which it should be, this being `ember-cli-rails` and all), then engines and railties come along for the ride, right?

If not, I'll gladly add it back to the gemspec

**Comment by rwz** [comment]
Also, what do you think about implicitly mounting the engine in development environment? But maybe use less obvious namespace, like `/ember-cli-tests/:app_name` instead? Then we could save some user's effort setting it up and not be a nuisance to those who don't care about this feature.

File: lib/ember-cli-rails.rb (hunk 1 above)
Comment (rwz) [inline lib/ember-cli-rails.rb]: `Rails` constant is defined in `railties`. So in this particular case we're depending on either `rails` or `railties`. I don't really have a hard preference on either of these, but one of them definitely should be there.

**Comment by seanpdoyle** [comment]
I am against implicitly mounting the engine.

What happens if people want to configure the URL?

We could add another configuration value:

```ruby
EmberCLI.configure do |config|
  config.tests_root = "/my-ember-cli-tests"
end
```

Or we could just depend on the pre-existing API provided by Rails engines.

For users who don't care about this feature, they're opted out by default, and don't have to worry about our obscure URL convention conflicting with their existing routes.

Also, requiring an explicit mount frees us from worrying about catch-all routes.

```ruby
# config/routes.rb

# this would have precedent over our route
match "*route-to-ember" => "ember#index"

# this would match before the catchall
mount EmberCLI::Engine => "tests"
match "*route-to-ember" => "ember#index"
```

**Comment by seanpdoyle** [comment]
It's worth noting that `<script src="testem.js">` is 404ing, since it isn't in the asset pipeline.

This doesn't seem to affect the test runnner, though

**Comment by rwz** [comment]
It seems to be an empty dummy file: https://github.com/ember-cli/ember-cli/blob/v0.1.15/lib/broccoli/testem.js

Seems like we could safely ignore the fact it's missing. The other option would be to do something like this:

```ruby
get 'testem', to: -> { [200, {}, [""]] }, format: :js
```

**Comment by rwz** [comment]
... or even better, handle it in the middleware. :)

**Comment by seanpdoyle** [comment]
For the time being, I'm ok with it 404ing.

**Comment by rwz** [comment]
Thanks again, you did an awesome job implementing this feature. :+1: :+1: :+1: 

Do you want to be a maintainer if the gem? You seem to care and put in a ton of work implementing features. We could definitely use all the help we could get from highly motivated people like you :)

**Comment by seanpdoyle** [comment]
@rwz I'd love to!

**Comment by rwz** [comment]
@seanpdoyle done.

**Comment by rondale-sc** [comment]
@seanpdoyle @rwz :+1: :beers:

**Pull request merged**